File: apps/cargo-tracker/app/cargotracker/billing/domain/model/valueobjects/DiscountRate.scala
package cargotracker.billing.domain.model.valueobjects

/** 割引率（0.0000〜0.3000 の範囲、4 桁小数）。 */
opaque type DiscountRate = BigDecimal

object DiscountRate:
  sealed trait Error
  case object OutOfRange extends Error

  private val Min: BigDecimal = BigDecimal("0.0000")
  private val Max: BigDecimal = BigDecimal("0.3000")

  def apply(value: BigDecimal): Either[Error, DiscountRate] =
    if value == null || value < Min || value > Max then Left(OutOfRange) else Right(value)

  def unsafeFrom(value: BigDecimal): DiscountRate = value
  def zero: DiscountRate = Min

  extension (r: DiscountRate)
    def value: BigDecimal = r
    def percent: BigDecimal = r * 100

/** 割引適用ポリシー種別。 */
enum DiscountPolicyType:
  case CorporateStandard
  case VolumeDiscount
  case Seasonal
  case None

File: docs/article/source/java-4/apps/cargo-tracker/backend/bookingms/src/main/java/com/example/cargotracker/booking/domain/model/valueobjects/DiscountRate.java
package com.example.cargotracker.booking.domain.model.valueobjects;

import com.example.cargotracker.shared.domain.error.BusinessRuleViolation;
import java.math.BigDecimal;

/** 法人契約の割引率。0.0000〜0.3000（domain-model.md）。 */
public record DiscountRate(BigDecimal value) {

    private static final BigDecimal MIN = new BigDecimal("0.0000");
    private static final BigDecimal MAX = new BigDecimal("0.3000");

    public DiscountRate {
        if (value == null) {
            throw new BusinessRuleViolation("割引率は必須です");
        }
        if (value.compareTo(MIN) < 0 || value.compareTo(MAX) > 0) {
            throw new BusinessRuleViolation("割引率は 0.0000〜0.3000 の範囲です: " + value);
        }
    }
}
